R/agg_helpers.R

```diff
@@ -0,0 +1,240 @@
+    if (col_type == "categorical") {
+      result_dt <- rbind(result_dt, aggregated_same_groupings_dt, use.names = T)
+    } else if (col_type == "interval") {
+      result_dt[[i]] <- aggregated_same_groupings_dt
+    }
+  }
+  if (col_type == "interval") {
+    result_dt <- rbindlist(result_dt, use.names = TRUE)
+    result_dt[, c(col_stem) := NULL]
+  }
+  return(result_dt)
+}
+
+#' @title Helper function to aggregate interval variables
+#'
+#' @description Helper function to aggregate interval variables by groups of data
+#'   that have the same intervals. Calls `agg_categorical()` for each group.
+#'
+#' @inheritParams agg
+#'
+#' @return Invisibly returns `dt` but throws error if inputs are not formatted
+#'   correctly.
+agg_interval <- function(dt,
+                         id_cols,
+                         value_cols,
+                         col_stem,
+                         col_type,
+                         mapping,
+                         agg_function = sum,
+                         missing_dt_severity = "stop",
+                         present_agg_severity = "stop",
+                         overlapping_dt_severity = "stop",
+                         na_value_severity = "stop",
+                         collapse_interval_cols = FALSE,
+                         quiet = FALSE) {
+
+  cols <- paste0(col_stem, "_", c("start", "end"))
+
+  # create one column to describe each interval
+  gen_name(mapping, col_stem = col_stem, format = "interval")
+  data.table::setnames(mapping, paste0(col_stem, "_name"), col_stem)
+
+  # create one column to describe each interval
+  dt_intervals <- unique(dt[, .SD, .SDcols = cols])
+  gen_name(dt_intervals, col_stem = col_stem, format = "interval")
+  data.table::setnames(dt_intervals, paste0(col_stem, "_name"), col_stem)
+  data.table::setkeyv(dt_intervals, cols)
+  dt <- dt[dt_intervals, on = cols, nomatch = 0]
+
+  # identify unique combinations of intervals to be aggregated
+  groups <- dt[, list(col = paste(get(col_stem), collapse = ",")), by = setdiff(names(dt), c(col_stem, cols, value_cols))]
+  unique_groups <- unique(groups$col)
+
+  all_groups_dt <- lapply(1:length(unique_groups), function(i) {
+    g <- unique_groups[i]
+    if (!quiet) message("Interval group ", i, " of ", length(unique_groups), ": ", g)
+
+    group_dt <- groups[col == g]
+    group_dt[, col := NULL]
+
+    # get rows of data that have the same combination of intervals
+    if (nrow(group_dt) == 0 & length(unique_groups) == 1) {
+      # when only one group exists and the only id variables are `col`
+      agg_dt <- dt
+    } else {
+      agg_dt <- dt[group_dt, on = names(group_dt), nomatch = 0]
+    }
+
+    dt_intervals <- unique(agg_dt[, .SD, .SDcols = cols])
+
+    # check for overlapping intervals
+    if (overlapping_dt_severity != "skip") {
+      overlapping_dt <- identify_overlapping_intervals(dt_intervals, identify_all_possible = overlapping_dt_severity != "none")
```

Comment:
with the changes to the overlap functions this will need to change?

Reply:
At least not how they are currently named, thats why I left these smaller helper functions with the same name, they just are no longer exported https://github.com/ihmeuw-demographics/hierarchyUtils/pull/77#discussion_r719807662